### ISSUE TITLE
vscode: in "while ((inputLine = in.readLine()) != null)"-loop. x = x + a without warning. Versus: x += a "value of the local variable is not used"

### ISSUE THREAD

**Issue by lorenzznerol** [redhat-developer/vscode-java#1518]
vscode: in "while ((inputLine = in.readLine()) != null)"-loop. x = x + a without warning. Versus: x += a "value of the local variable is not used"

This is a full copy of

> https://stackoverflow.com/questions/62854005/vscode-in-while-inputline-in-readline-null-loop-x-x-a-without

One of the answers was that it is probably a vscode issue. It could also be an issue of AdoptOpenJDK or a genuine Java compiler issue, but that is less likely.

##### Environment
- Operating System: Windows_NT ia32 10.0.18363
- JDK version: AdoptOpenJDK JDK with Hotspot 14.0.1.7 (x86)
- Java extension version: Name: Java Extension Pack Version: 0.9.1
- Visual Studio Code version: 1.47.0 (user setup)
- Commit: d5e9aa0227e057a60c82568bf31c04730dc15dcd
- Date: 2020-07-09T08:01:16.801Z
- Electron: 7.3.2
- Chrome: 78.0.3904.130
- Node.js: 12.8.1
- V8: 7.8.279.23-electron.0

##### Steps To Reproduce
The following is an example from https://docs.oracle.com/javase/tutorial/networking/urls/readingURL.html, the only change is that the content of the String inputLine is added up to the strAll String instead of printing it, see X_HERE.

    import java.net.*;

    import java.io.*;

    

    public class URLReader {

        public static void main(String[] args) throws Exception {

    

            URL oracle = new URL("http://www.oracle.com/");

            BufferedReader in = new BufferedReader(

            new InputStreamReader(oracle.openStream()));

    

            String inputLine;

            String strAll = "";

            while ((inputLine = in.readLine()) != null)

                // System.out.println(inputLine);

                strAll = strAll + inputLine; // X_HERE

            in.close();

        }

    }

This does not throw a warning in the "Problems" window of vscode. If you replace line X_HERE with 

    

    strAll += inputLine;

using "addition assignment" you get the warning: 

> The value of the local variable strAll is not used

If you replace the while condition with (true), the warning for strAll disappears. Why is "addition assignment" treated differently in this context?

**Comment by fbricon** [comment]
It's the same thing in Eclipse IDE. It could be a bug of the ECJ compiler, or a "feature". Maybe @stephan-herrmann or @rgrunber might be able to tell.

**Comment by stephan-herrmann** [comment]
`strAll += inputLine` is a "compound assignment expression". Because this is a built-in structure of Java, ecj knows about the specifics of this construct, in particular it tracks that the implicit read doesn't create a data flow from `strAll` to another variable or argument.

`strAll = strAll + inputLine`, however is just one out of a myriad of forms, where the human reader will readily see that `strAll` really flows no-where.

While the decision to treat both forms differently was made before my time in JDT, I am pretty sure that the team saw this as the cleanest line to be drawn between things ecj can reason about and things that are just too complex. Some examples of what would need to be discussed when extending the flow analysis:

* `strAll = "foo" + strAll + "bar";`
* `strAll = (bar = strAll) + "foo";`
* `strAll = convert(strAll);` _// this one is trickier than it looks on first sight_
*  etc. pp.

So the pattern would likely be that JDT gives you a finger and people will take the whole arm, or more ...

Bottom line: Use Java specifics appropriately if you want to give more information to the Java compiler :)


Regarding `while (true)`, I _think_ (without looking closer), that flow analysis just stops short when it finds that a method cannot complete normally.

**Comment by fbricon** [comment]
@stephan-herrmann thanks for the detailed analysis, as always :-)

**Comment by lorenzznerol** [comment]
@stephan-herrmann just making sure you got it right: `strAll = strAll + inputLine;` works, `strAll += inputLine;` does not. From your answer here I read the opposite. I might be wrong.

**Comment by stephan-herrmann** [comment]
@lorenzznerol  Define "works" :)

Are you saying you don't want to see a warning?

If so, please tell me which part of the observable behavior changes when you replace `strAll += inputLine;` by `;` and then remove the variable `strAll` as the warning implicitly suggests?

**Comment by lorenzznerol** [comment]
@stephan-herrmann Yes it was about not to get the warning that is shown when using `strAll += inputLine;`. And I am afraid I cannot test it now when not using `strAll` at all, but I guess it will throw the same warning. Good idea to check that, though I think I had done that before raising the issue. 

I do not have a Java environment installed now and will not do it only to test this here. It is not really *my issue* but rather a hint for the developers who might be interested.

**Comment by stephan-herrmann** [comment]
> Yes, the aim was not to get the warning that is shown when using strAll += inputLine;. 

@lorenzznerol  That's where we disagree. The warning (where it is given) is good, as also confirmed by the accepted answer on StackOverflow. That same answer only expressed a surprise that the warning is given _only_ in one case, not the other. The reason behind that is what I explained in my first comment here.

**Comment by lorenzznerol** [comment]
@stephan-herrmann got it. The warning is good, what a convincing idea, I withdraw my recent remarks. Strangely, I must have forgotten this core idea. It was just too long ago, I should have read it again.

**Comment by stephan-herrmann** [comment]
> The warning is good

Thanks for confirming :)

**Comment by testforstephen** [comment]
@lorenzznerol Can we close this issue as "Won't fix"?

**Comment by lorenzznerol** [comment]
Yes
